### PR TITLE
feat(generic-child-tree): Implement generic key decoding and generic …

### LINKED ISSUE #1042: [private graph] More generic child tree utility
### Description
Currently we have a child tree utility that allow us to interact with a child tree but it lacks some generic items such as
- Generic hash function
- Generic Key type

Not having a generic key type forces us to decode all keys to `Vec<u8>` first and then we can decode to the desired key type.

### AC
- Support Generic Reversible hash function
- Support Generic Key Type

### Additional info
for an example of how this can be done we can look at https://github.com/paritytech/substrate/blob/38c3e5ee10be760115a77a115ef0ac6ba381542b/frame/support/src/storage/generator/double_map.rs#L76

### PR DESCRIPTION
**Pull request by JoeCap08055**
# Goal
The goal of this PR is to implement generics for child tree key types and hash method.

Closes #1042 

# Checklist
- [ ] Chain spec updated
- [ ] Custom RPC OR Runtime API added/changed? Updated js/api-augment.
- [ ] Design doc(s) updated
- [ ] Tests added
- [ ] Benchmarks added
- [ ] Weights updated

### CODE REVIEW

File: pallets/stateful-storage/src/lib.rs
Comment (saraswatpuneet) [inline pallets/stateful-storage/src/lib.rs:85]: :1st_place_medal:

Comment (aramikm) [reply pallets/stateful-storage/src/lib.rs:85]: There might be a miscommunication on my part. We want `TwoX` for this pallet but for tests and maybe other pallets the `Identity` default makes sense

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (aramikm) [inline pallets/stateful-storage/src/stateful_child_tree.rs:51]: I don't think this is going to work. We need to hash each key separately and concat those hashes. To be able to look them up by prefix

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:85]: Right; I was waiting for a decision before updating this. I'll change to `TwoX`.

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/stateful_child_tree.rs:51]: Ah, right. Duh. `hash(x) + hash(y) <> hash(x + y)`

File: pallets/stateful-storage/src/types.rs
Comment (aramikm) [inline pallets/stateful-storage/src/types.rs:8]: It seems like you have duplicated these types in lib.rs also. I would suggest to define all these types inside here and remove the duplicates defined in `lib.rs`

**Review by aramikm — APPROVED** [review]
Looks great!. Just need to rebase from `private-graph-main` since my changes got in.

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:181]: Can we please add some docs for this type,

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/stateful_child_tree.rs:40]: This trait does more than check if something is a stateful key. I don't have any suggestions for naming at the moment.

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:324]: Can we call it like this?
StatefulChildTree::<T::Hasher>::try_read::<_, ...>(&msa_id, &key). unwrap_or_default()

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/stateful_child_tree.rs:248]: Do you expect that different hashers are going to be used on the same type? I guess why not set this as a generic one level up?

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/stateful_child_tree.rs:260]: the match thing seems more of a rust way than an if statement. No strong opinion.

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/stateful_child_tree.rs:24]: nice

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:181]: Removed; after merge from pg-main, got `ItemizedKey` from `common_primitives::stateful_storage`

File: pallets/stateful-storage/src/types.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/types.rs:8]: Done.

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/stateful_child_tree.rs:40]: Hmm, good point. It kind of grew. Maybe `MultipartKey` is a bettername?

Comment (JoeCap08055) [reply pallets/stateful-storage/src/stateful_child_tree.rs]: Renamed to `MultipartKey`

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/stateful_child_tree.rs]: I'm not understanding what you're getting at... the hasher is a type parameter to `StatefulChildTree`, but it needs to make sure that the MultipartKey uses the same hasher. What am I missing?

File: pallets/stateful-storage/src/stateful_child_tree.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/stateful_child_tree.rs:260]: It's more than an `if`... it's an `if let`. Which is just syntactic sugar for a basic `match`.

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs]: No... the return value is an `Option` wrapped in a `Result`. `unwrap_or_default()` on the Result may still yield a `None` value.

**Pull request merged**
